R/prepare_historical_xls.R

```diff
@@ -0,0 +1,176 @@
+    mutate_all(~ifelse(is.na(.), 0, .)) %>% #change NAs to 0
+    hist_xls_change_xs_to_ones() %>% 
+    #ensure counts are numeric
+    mutate_at(c("s1","s2","s3","s4","s5","s6","s7","s8","s9","s10","s11","s12","s13","s14","s15","s16","s17","s18","s19", "s20"), as.integer) %>%
+    hist_xls_match_mbbs_format()
+  
+  return(hist_xls)
+}
+
+#'Filters to only rows where the first column is a species code
+#'that is four capital letters
+#'or acsp - "Accipiter sp."
+#'@importFrom stringr str_detect
+#'@importFrom dplyr filter
+#'@param x a historical .xls 
+hist_xls_filter_to_species_code_rows <- \(x){
+  x %>%
+    filter(str_detect(.[[1]], "^[A-Z][A-Z][A-Z][A-Z]$|acsp"))
+}
+
+
+#' Rename the columns of a historical xls
+#'@importFrom dplyr %>%
+#'@param x a historical .xls 
+hist_xls_rename_columns <- \(x){
+  x %>%
+    `colnames<-`(c("species_code", "s1","s2","s3","s4","s5","s6","s7","s8","s9","s10","s11","s12","s13","s14","s15","s16","s17","s18","s19", "s20", "sum", "sequence","species_code2", "common_name"))
+}
+
+#' Flags if a species code has been missed
+#' @param df a dataframe that contains the column "sequence"
+hist_xls_flag_missed_species <- function(df){
+  
+  if(nrow(df) != max(df$sequence)) {
+    return("Error! nrows does not match expected number of species codes")
+  } 
+  
+  return(df)
+}
+
+
+#' Pull mbbs_county, route_num, and year
+#' out of a historical MBBS_XL
+#' route_num MUST BE the first number in the filename
+#' year MUST BE the second number in the filename
+#' @importFrom stringr str_extract str_extract_all
+#' @param filename a filename that contains the county, route num, and year
+hist_xls_extract_county_num_year_from_filename <- function(filename) {
+  
+  mbbs_county <- str_extract(filename, "[cC]hatham|[dD]urham|[oO]range") %>%
+    tolower()
+  
+  nums <- str_extract_all(filename, "[0-9]+")[[1]]
+  route_num <- nums[1] %>% 
+    as.integer()
+  year <- nums[2] %>%
+    as.integer()
+  
+  survey_information <- data.frame(mbbs_county, route_num, year)
+  return(survey_information)
+}
+
+
+#'Converts 'x' and 'X' counts to 1
+#'following method of mbbs route information
+#'as compiled by Haven Wiley
+#'@importFrom purrr map_dfr
+#'@importFrom dplyr %>%
+#'@param hist_xls a historical mbbs excel file
+hist_xls_change_xs_to_ones <- function(hist_xls) {
+  hist_xls <- hist_xls %>%
+  purrr::map_dfr(., ~ifelse(.=="x",1,.)) %>%
+  purrr::map_dfr(., ~ifelse(.=="X",1,.))
+}
+
+
+
+#' Match MMBS format
+#' This doesn't add every mbbs column
+#' But fills in those for species or route info
+#' that are expected
+#' @importFrom dplyr left_join select
+#' @param hist_xls a historical mbbs excel file
+#' @param mbbs_survey_events rda of all surveys that have been run on the mbbs
+hist_xls_match_mbbs_format <- function(hist_xls, mbbs_survey_events = "data/mbbs_survey_events.rda"){
+  
+  #bring in other dfs
+  taxonomy <- get_ebird_taxonomy()
+  load(mbbs_survey_events)
```

Comment:
A call to `load` here smells funny to me, though I don't have another thought off the top of my head.

Reply:
does load("data/mbbs_survey_events.rda") improve it?

Reply:
No, it's using `load` period that bugs me. But it's fine for now.

FWIW - I'm fine with `load` in scripts but in functions, I prefer to avoid this kind of call with side-effects.

Reply:
data("mbbs_survey_events")?

Reply:
No, same problem in terms of side effects.